### PR TITLE
handle combined participant update.

### PR DESCRIPTION
**Pull request by cloudwebrtc**
https://github.com/livekit/client-sdk-js/pull/275

### CODE REVIEW

File: test/core/room_e2e_test.dart [hunk 1]
```diff
@@ -36,10 +36,16 @@ void main() {
     test('participant join', () async {
       expect(
         room.events.streamCtrl.stream,
-        emits(predicate<ParticipantConnectedEvent>(
-          (event) => event.participant.sid == remoteParticipantData.sid,
-        )),
+        emitsInOrder(<Matcher>[
```
Comment (davidzhao) [inline test/core/room_e2e_test.dart:39]: is this checking to see TrackPublishedEvent is emitted before ParticipantConnectedEvent? I think that's the wrong behavior.

File: lib/src/participant/remote.dart [hunk 2]
```diff
@@ -167,15 +166,13 @@ class RemoteParticipant extends Participant<RemoteTrackPublication> {
...
-        );
-        [events, room.events].emit(event);
-      }
+    // always emit events for new publications, Room will not forward them unless it's ready
+    for (final pub in newPubs) {
+      final event = TrackPublishedEvent(
+        participant: this,
+        publication: pub,
+      );
+      [events, room.events].emit(event);
```
Comment (davidzhao) [inline lib/src/participant/remote.dart:175]: I don't believe we want to do this here. TrackPublished events should not be emitted before the Room is connected.

When a room is already connected
* fire `TrackPublished`

When a room is not yet connected
* do *not* fire `TrackPublished`

**Review by davidzhao — CHANGES_REQUESTED** [review]

File: lib/src/participant/remote.dart (hunk 2 above)
Comment (cloudwebrtc) [reply lib/src/participant/remote.dart:175]: If the room here is not connected, do we need to cache the TrackPublished event until room is reconncted and then emit it?

Comment (davidzhao) [reply lib/src/participant/remote.dart:175]: I don't think we need to cache it

Comment (davidzhao) [reply lib/src/participant/remote.dart:175]: the event is only fired for tracks that are published after the participant has joined

Comment (cloudwebrtc) [reply lib/src/participant/remote.dart:175]: so, here we need to confirm that the `subscriber.pc` is connected, then the event can be emitted, is this correct?

File: lib/src/core/signal_client.dart [hunk 3]
```diff
@@ -240,6 +240,13 @@ class SignalClient extends Disposable with EventsEmittable<SignalEvent> {
       _updateConnectionState(ConnectionState.disconnected);
     }
   }
+
+  @internal
+  void emitWhenConnected(SignalEvent event) {
```
[the remark this answers is not in the document]
Comment (cloudwebrtc) [reply lib/src/core/signal_client.dart:245]: okay, I'll move the code to the correct location.

File: lib/src/participant/remote.dart (hunk 2 above)
Comment (davidzhao) [reply lib/src/participant/remote.dart:175]: not exactly.. to be clear, when I say "connected", it's referring to Room.state == connected. This is the contract we have with developers

Comment (cloudwebrtc) [reply lib/src/participant/remote.dart:175]: ok, I'm trying to figure out the timing,  there are two `_updateConnectionState(ConnectionState.connected);` in core/engine.dart that let me know the room is connected. 
so, no more questions

**Review by davidzhao — APPROVED** [review]

**Pull request merged**